### ISSUE TITLE
Confusion Matrix doesn't show absent labels

### ISSUE THREAD

**Issue by jerpint-mila** [comet-ml/issue-tracking#535]
## Describe the Bug
I want to log a confusion matrix where instances of a label are not seen in the test set. This can happen e.g. if I train on a dataset with a very large training set with diverse categories, and evaluate on a subsample of the data where some of the infrequent classes are not present. When this happens, I would still like to see an empty row/column for that class in the confusion matrix. Currently, the API silently fails.

Note that producing the confusion matrix with absent classes is very much compatible with `sklearn.metrics.confusion_matrix` yet still fails to get uploaded using the comet python SDK (see in example below).

## Expected behavior
I would like to see the "absent" labels with row/column filled with zeros.

## Where is the issue?
- [x] Comet Python SDK
- [ ] Comet UI
- [ ] Third Party Integrations (Huggingface, TensorboardX, Pytorch Lighting etc)

## To Reproduce
Here is exact code to reproduce the behaviour:

```python
import os
import comet_ml
from sklearn.metrics import confusion_matrix

experiment = comet_ml.Experiment(
    api_key=os.getenv("COMET_API_KEY"),
    project_name="confusion-matrix-test"
)


# I have 7 labels in my experiments, but not in all subsets of data (i.e. some labels might be seen during train but not test)
labels = ["a", "b", "c", "d", "e", "f", "g"]
assert len(labels) == 7

# model predictions (at train time)
y_true = [1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 2, 4, 3, 1, 1, 2]
y_pred = [0, 0, 2, 3, 1, 0, 3, 0, 4, 0, 0, 0, 0, 6, 6, 5]

# Logs as expected
experiment.log_confusion_matrix(
    y_true=y_true,
    y_predicted=y_pred,
    labels=labels,
    title="train confusion matrix",
    file_name="train-confusion-matrix.json",
)

# model predictions (at test time), notice that class [5,6] are NOT present in this subset
y_true = [1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 2, 4, 3, 1, 1, 2]
y_pred = [0, 0, 2, 3, 1, 0, 3, 0, 4, 0, 0, 0, 0, 0, 0, 0]

# Fails to log with a simple "ignoring" message
experiment.log_confusion_matrix(
    y_true=y_true,
    y_predicted=y_pred,
    labels=labels,
    title="test confusion matrix",
    file_name="test-confusion-matrix.json",
)

# Get a list of all present classes and labels in preds and targets
present_classes = list(set(y_pred).union(set(y_true)))
present_labels = [labels[cls] for cls in present_classes]
assert len(present_labels) == 5  # classes 5, 6 (e, f) not present in test set

# Only after filtering "absent" labels, it succeeds
# Succeeds
experiment.log_confusion_matrix(
    y_true=y_true,
    y_predicted=y_pred,
    labels=present_labels,
    title=f"test confusion matrix",
    file_name=f"test-confusion-matrix.json",
)


# Alternative - still fails
# Compute conf. matrix using sklearn
conf_mat = confusion_matrix(y_true=y_true, y_pred=y_pred, labels=range(len(labels)))
print(conf_mat)

# Fails
experiment.log_confusion_matrix(
    matix=conf_mat,
    labels=labels,
    title="test confusion matrix",
    file_name="test-confusion-matrix.json",
)

```

## Stack Trace
If possible please include the full stack trace of your issue here

```
COMET ERROR: Error logging confusion matrix; ignoring
COMET ERROR: Attempt to log empty confusion matrix; ignoring
```

## Comet Debug Log
If possible, please follow the instructions [here](https://www.comet.ml/docs/python-sdk/advanced/#troubleshooting) to run Comet in debug mode and attach the resulting log file.  

## Screenshots or GIFs
<img width="1144" alt="image" src="https://github.com/comet-ml/issue-tracking/assets/106993305/212fb1de-4127-4b37-abf5-880d75563244">

In the screenshot, you see the "suceeded" upload of the test confusion matrix, but doesn't show the classes that aren't present. I would like to "see" that there are classes that are simply not being predicted from the confusion matrix.

## Link to Comet Project/Experiment
If applicable, please provide a link to your Comet Project or Experiment.

**Comment by jerpint-mila** [comment]
Here's another interesting behaviour: if we have missing classes, but a present class much higher, it will then consider those classes and show the zero'd rows/columns, which is the "expected" behaviour in that case (inlcuding assuming class 0 to be present). So it seems like you only can't append new classes it hasn't seen.

```python
import os
import comet_ml
from sklearn.metrics import confusion_matrix

experiment = comet_ml.Experiment(
    api_key=os.getenv("COMET_API_KEY"),
    project_name="confusion-matrix-test"
)


# I have 7 labels in my experiments, but not in all subsets of data (i.e. some labels might be seen during train but not test)
labels = ["a", "b", "c", "d", "e", "f", "g"]
assert len(labels) == 7

# model predictions (at train time), note that class 0 is not present in this set, nor is 3, 4
y_true = [1, 1, 1, 1, 5]
y_pred = [1, 2, 2, 1, 5]

# Get a list of all present classes and labels in preds and targets
present_classes = list(set(y_pred).union(set(y_true)))
present_labels = [labels[cls] for cls in present_classes]

# Fails, because it assumes class 0 to be part of our data, 3, 4
experiment.log_confusion_matrix(
    y_true=y_true,
    y_predicted=y_pred,
    labels=present_labels,
    title=f"test confusion matrix",
    file_name=f"test-confusion-matrix.json",
)

# Get a list of all present classes and labels in preds and targets, include labels in range max(classes)
present_classes = list(set(y_pred).union(set(y_true)))
present_labels = [labels[cls] for cls in range(max(present_classes) + 1)]

# Succeeeds, because now we included up to 5 in the labels
experiment.log_confusion_matrix(
    y_true=y_true,
    y_predicted=y_pred,
    labels=present_labels,
    title=f"test confusion matrix",
    file_name=f"test-confusion-matrix.json",
)
```

Here, it will 

<img width="612" alt="image" src="https://github.com/comet-ml/issue-tracking/assets/106993305/c49f51b4-33b4-4d6f-979b-d01d8f15b91f">

**Comment by dsblank** [comment]
Thank you for the detailed analysis! Taking a look...

**Comment by dsblank** [comment]
You indeed find a bug, so thank you for that.

You have a bug on 4th one in the first code: "matix" should of course be "matrix".

Can you check out these resulting confusion matrices after fixing the bug to see if they look ok?

https://www.comet.com/dsblank/confusion-matrix-test/c832109ef29144798c3913b2a8b5256d?assetId=5e53ddcb6ac040258d160104809748d5,86be638481b549cb8bc148c0038f54dc,2bb25244696b48aa8db821791c14c35a,5ebcc34561494203b50615da4067ec0f&cellValue=Counts&colorDistribution=Equal&experiment-tab=confusionMatrix

**Comment by jerpint-mila** [comment]
Ah yes, must have been a typo from copy-pasting from elsewhere.

I am guessing that the new matrix should be the one in `test-confusion-matrix4.json`? If so looks correct to me, thanks for looking at that so quickly

**Comment by dsblank** [comment]
No problem. The 1 - 4 are the examples in your first code. They all look correct?

**Comment by jerpint-mila** [comment]
In that case, I would expect `test-confusion-matrix2.json` to be identical to `test-confusion-matrix4.json`, with both displaying all instances of the labels with the empty rows/columns for the labels that are not present

**Comment by dsblank** [comment]
Ok, thanks. We'll see if we can fix that too.

**Comment by dsblank** [comment]
Updated results: https://www.comet.com/dsblank/confusion-matrix-test/736dbc8531f34aea8b80ba73af1715e1?assetId=85febf119d0f436c9132cb26e9e00ba6,6792130b34dd4e61943092edf0a14ef8,38beb254915f4c54b2b64d45705df952,219b216b089c40c1ad6c85688d5c9f91&cellValue=Counts&colorDistribution=Equal&experiment-tab=confusionMatrix&viewId=4wdIUhSm92KlfJJcGlEpzQYs1

Tracking as internal issue CM-9571.

**Comment by dsblank** [comment]
Fix will probably be released this week.

**Comment by dsblank** [comment]
Please update to comet_ml version 3.37.2 and see if that fixes your issues.
